python/extensions/python.bzl

```diff
@@ -60,82 +60,97 @@ def _python_register_toolchains(toolchain_attr, version_constraint):
     )
 
 def _python_impl(module_ctx):
-    # Use to store all of the toolchains
+    # The toolchain info structs to register, in the order to register them in.
     toolchains = []
 
-    # Used to check if toolchains already exist
-    toolchain_names = []
-
-    # Used to store toolchains that are in sub modules.
-    sub_toolchains_map = {}
+    # We store the default toolchain separately to ensure it is the last
+    # toolchain added to toolchains.
     default_toolchain = None
-    python_versions = {}
+
+    # Map of toolchain name to registering module
+    global_toolchain_names = {}
+
+    # Map of string Major.Minor to the toolchain name and module name
+    global_toolchain_versions = {}
 
     for mod in module_ctx.modules:
+        module_toolchain_names = []
+        module_toolchain_versions = []
+
         for toolchain_attr in mod.tags.toolchain:
-            # If we are in the root module we always register the toolchain.
-            # We wait to register the default toolchain till the end.
+            toolchain_name = toolchain_attr.name
+
+            # Duplicate names within a module indicate a misconfigured module.
+            if toolchain_name in module_toolchain_names:
+                _fail_duplicate_module_toolchain_name(mod.name, toolchain_name)
+            module_toolchain_names.append(toolchain_name)
+
+            # Ignore name collisions in the global scope because there isn't
+            # much else that can be done. Modules don't know and can't control
+            # what other modules do, so the first in the dependency graph wins.
+            if toolchain_name in global_toolchain_names:
+                _warn_duplicate_global_toolchain_name(
+                    toolchain_name,
```

Comment:
For toolchains, they don't really need a user-provided name, so shouldn't have one. More generally speaking, it's not safe/reliable for a submodule to request a toolchain with settings that can't be encoded in the toolchain's constraint settings. The args for enabling coverage or ignoring the root user-thing are examples of this. (I think there might be one more, but can't recall atm).

This is why I keep beating the "remove the name arg (and others args)" drum as side-comments in PRs :). In rules_rust, for example, they don't have a name arg. In rules_go, they restrict who can use the name arg. We should follow suit, too (ignoring it works well enough for now).

For something like pip, it's a looming problem, because each pip resolution is likely unique. Fabian has a draft PR implementing an "isolated extension" feature, where you can mark an extension as "isolated", and then all the repos it creates are in its own namespace.

Reply:
Unless you can figure out some bazel magic we will need a name for the multiple python versions. Maybe we write out everything to a repo that has a standard name, rather than have different repos for each python version 🤷🏻‍♂️ we need to access :defs.bzl for the different python versions.

Reply:
To clarify: It's just the user-selection of the repo names that is problematic. Our current repo layout for the toolchains is OK, and the defacto naming convention ("python_x_y") is also fine.